clash-lib/src/Clash/Driver/BrokenGhcs.hs

```diff
@@ -0,0 +1,140 @@
+
+module Clash.Driver.BrokenGhcs where
+
+import Data.Maybe (listToMaybe)
+import Data.Version (Version(Version, versionBranch))
+import GHC.Platform (OS(..))
+import System.Info (fullCompilerVersion)
+
+import qualified Clash.Util.Interpolate as I
+import qualified System.Info
+
+-- | Current OS. Currently only recognizes Linux, Windows, and macOS.
+os :: OS
+os = case System.Info.os of
+  "darwin" -> OSDarwin
+  "linux" -> OSLinux
+  "mingw32" -> OSMinGW32
+  _ -> OSUnknown
+
+-- | What OS GHC is broken on (or all)
+data BrokenOn = All | SomeOs OS
+
+data GhcVersion
+  = GhcMajor Int Int
+  -- ^ @A.B.x@, for example @8.10.x@, where @x@ can be any number
+  | GhcMinor Int Int Int
+  -- ^ @A.B.C@, for example @9.10.2@
+
+data Why = Why
+  { what :: String
+    -- ^ What is broken
+  , solution :: String
+    -- ^ What can be done to work around or solve the issue
+  , issue :: String
+    -- ^ Link to issue
+  , brokenOn :: [(BrokenOn, GhcVersion)]
+    -- ^ What operation systems are affected
+  }
+
+-- | Get current GHC version expressed as a triple. It probably does something
+-- non-sensible on unreleased GHCs.
+ghcVersion :: (Int, Int, Int)
+ghcVersion =
+  case fullCompilerVersion of
+    Version{versionBranch} ->
+      case versionBranch of
+        [] -> (0, 0, 1)
+        [a] -> (a, 0, 1)
+        [a, b] -> (a, b, 1)
+        [a, b, c] -> (a, b, c)
+        (a:b:c:_) -> (a, b, c)
+
+ghcMajor, ghcMinor, ghcPatch :: Int
+(ghcMajor, ghcMinor, ghcPatch) = ghcVersion
+
+-- | Pretty print 'Why' into an error message
+whyPp :: Why -> String
+whyPp Why{what, solution, issue}= [I.i|
+  Clash has known issues on #{ghcMajor}.#{ghcMinor}.#{ghcPatch} on your current
+  OS. While not completely preventing the compiler from working, we recommend
+  switching to another GHC version. Symptoms:
+
+    #{what}
+
+  Consider the following work around or solution:
+
+    #{solution}
+
+  More information can be found at:
+
+    #{issue}
+
+  If you want to ignore this message, pass the following flag to Clash:
+
+    -fclash-ignore-broken-ghcs
+  |]
+
+-- | Which GHCs are broken and why
+brokenGhcs :: [Why]
+brokenGhcs = [brokenClashCores, brokenTypeErrors, slowStarts]
+ where
+  brokenClashCores = Why
+    { what = "GHC fails to compile clash-cores"
+    , solution = "Upgrade to GHC 9.4 or downgrade to GHC 9.0"
+    , issue = "<no link>"
+    , brokenOn = [(SomeOs OSMinGW32, GhcMajor 9 0)]
+    }
+
+  brokenTypeErrors = Why
+    { what = "Clash type error messages are indecipherable"
+    , solution = "Upgrade to GHC 9.4 or downgrade to GHC 9.0"
+    , issue = "<no link>"
+    , brokenOn = [(All, GhcMajor 9 2)]
+    }
+
+  slowStarts = Why
+    { what = "Clash starts really slowly from GHC 9.4.8 up to and including 9.6.2"
+    , solution = "Downgrade to GHC 9.4.7 or upgrade to GHC 9.6.3 or newer"
+    , issue = "https://github.com/clash-lang/clash-compiler/issues/2710"
+    , brokenOn = (All, GhcMinor 9 4 8) : [(All, GhcMinor 9 6 n) | n <- [1, 2]]
```

Comment:
Instead of listing all broken (minor) versions individually, I think it would be better to specify ranges of broken versions.
Something like:
```haskell
   , brokenOn = [(All, GhcRange { from = GhcMinor 9 4 8, to = GhcMinor 9 6 2 }]
```

Reply:
Done!